presto-main/src/main/java/com/facebook/presto/execution/scheduler/SourcePartitionedScheduler.java

```diff
@@ -372,8 +372,15 @@ else if (pendingSplits.isEmpty()) {
             return new ScheduleResult(false, overallNewTasks.build(), overallSplitAssignmentCount);
         }
 
-        // Only try to finalize task creation when scheduling would block
-        overallNewTasks.addAll(finalizeTaskCreationIfNecessary());
+        if (anyBlockedOnPlacements) {
+            // This prevents potential deadlock on the probe side of broadcast join,
+            // when both probe and builder size are large.
+            //
+            // When a deadlock happens, output buffer on the builder source stage cannot be purged before probe
+            // side task schedule finished. In the meantime the probe side split cannot be consumed
+            // since builder side hash table construction is not finished.
```

Comment:
This is how I would write it. Feel free to make changes as you see fit

```
// In a broadcast join, output buffers of the tasks in build source stage have to
// hold onto all data produced, even if the data is acknowledged by all known
// consumers.  It is necessary to hold onto the data because new consumers may be
// added until the probe side task scheduling finishes.  As a result, the build
// side may block due to a full output buffer.  In the meantime the probe side
// split cannot be consumed since builder side hash table construction has not
// finished.
//
// This leads to a deadlock because neither build nor probe can make any progress.